internal/llrp/device_service.go

```diff
@@ -295,3 +302,39 @@ func (ds DSClient) put(path string, data []byte) error {
 
 	return nil
 }
+
+// tryGet attempts to make an HTTP GET call to the device service at the specified path. It will
+// try up to maxTries times and sleeps sleepInterval*tryCount in-between tries. It will return the
+// raw response and error objects of the final HTTP call that was completed.
+// It determines a successful try as anything returning 2xx http code.
+// Note: if err is nil, caller is expected to close response body
+func (ds DSClient) tryGet(path string) (resp *http.Response, err error) {
+	req, err := http.NewRequest(http.MethodGet, ds.baseURL+path, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating new http GET request for path %s", path)
+	}
+
+	// start at 1 since this is our first try
+	for i := 1; ; i++ {
+		resp, err = ds.httpClient.Do(req)
+		if err != nil {
+			ds.lc.Error(fmt.Sprintf("device service HTTP GET %s attempt returned an error: %v", path, err))
+			if i < maxTries { // if we have tries left, sleep and retry
+				time.Sleep(sleepInterval * time.Duration(i))
+				continue
+			}
+		} else if !(200 <= resp.StatusCode && resp.StatusCode < 300) {
+			ds.lc.Error(fmt.Sprintf("device service HTTP GET %s attempt returned http error code %d", path, resp.StatusCode))
+			if i < maxTries { // if we have tries left, sleep and retry
+				_ = resp.Body.Close() // close the body so that the request may be re-used
+				time.Sleep(sleepInterval * time.Duration(i))
+				continue
+			}
+		}
+		// if we reached here that means either the request was successful, or
+		// we have no more tries left, so exit loop
+		break
+	}
+
+	return resp, err
```

Comment:
I think this might be more clear if you duplicate the `httpClient.Do`/`err` and `StatusCode` checks. I believe this is equivalent, but I've only edited it in the Github PR textbox. Either way, if you prefer the original style, it looks like it gets the job done, so it's fine by me.

```suggestion
      resp, err := ds.httpClient.Do(req)
      success := err == nil && (200 <= resp.StatusCode && resp.StatusCode < 300)
	for i := maxTries; i > 0 && !success; i-- {
	        if err != nil {
			ds.lc.Error(fmt.Sprintf("device service HTTP GET %s attempt returned an error: %v", path, err))
		} else {
			ds.lc.Error(fmt.Sprintf("device service HTTP GET %s attempt returned http error code %d", path, resp.StatusCode))
			_ = resp.Body.Close() // close the body so that the request may be re-used
		}
		
              time.Sleep(sleepInterval * time.Duration(i))
		resp, err = ds.httpClient.Do(req)
		success = err == nil && (200 <= resp.StatusCode && resp.StatusCode < 300)
	}

	return resp, err
```